File: lv1_module2_student/ros2_ws/src/turtle_cpp/src/distance_subscriber.cpp
#include <memory>
#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/float32.hpp"

class DistanceSubscriber : public rclcpp::Node
{
public:
  DistanceSubscriber()
  : Node("cpp_distance_subscriber")
  {
    subscription_ = create_subscription<std_msgs::msg::Float32>(
      "/turtle_distance", 10, [this](std_msgs::msg::Float32::SharedPtr msg) {
        RCLCPP_INFO(get_logger(), "distance: %.3f", msg->data);
      });
  }

private:
  rclcpp::Subscription<std_msgs::msg::Float32>::SharedPtr subscription_;
};

int main(int argc, char ** argv)
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<DistanceSubscriber>());
  rclcpp::shutdown();
  return 0;
}

File: lv1_module2/ros2_ws/src/turtle_py/turtle_py/distance_subscriber.py
"""Warn when the turtle is farther than a configurable threshold."""

import rclpy
from rclpy.node import Node
from rclpy.executors import ExternalShutdownException
from rclpy.qos import qos_profile_sensor_data
from std_msgs.msg import Float32


class DistanceSubscriber(Node):
    def __init__(self):
        super().__init__('distance_subscriber')
        self.declare_parameter('warn_distance', 2.5)
        self.declare_parameter('best_effort', False)
        qos = (qos_profile_sensor_data
               if self.get_parameter('best_effort').value else 10)
        self.create_subscription(Float32, '/turtle_distance', self.callback, qos)

    def callback(self, msg):
        threshold = self.get_parameter('warn_distance').value
        if msg.data > threshold:
            self.get_logger().warning(
                f'distance {msg.data:.3f} exceeds {threshold:.3f}')


def main(args=None):
    rclpy.init(args=args)
    node = DistanceSubscriber()
    try:
        rclpy.spin(node)
    except (KeyboardInterrupt, ExternalShutdownException):
        pass
    finally:
        node.destroy_node()
        if rclpy.ok():
            rclpy.shutdown()
